File: bridges/dotnet/src/Flint.Protocol/Framing.cs
using System;
using System.IO;
using Google.Protobuf;
using Flint.Protocol.V1;

namespace Flint.Protocol
{
    /// <summary>Blocking stream framing. The caller owns deadlines and closes on errors.</summary>
    public static class Framing
    {
        public const uint ProtocolVersion = 1;
        public const int MaxFrameBytes = 100 * 1024 * 1024;

        public static void Validate(Envelope envelope)
        {
            if (envelope.ProtocolVersion != ProtocolVersion) throw new InvalidDataException("unsupported protocol version");
            if (envelope.RequestId.Length == 0) throw new InvalidDataException("missing request_id");
            if (envelope.PayloadCase == Envelope.PayloadOneofCase.None) throw new InvalidDataException("missing or unsupported payload");
        }

        public static byte[] Encode(Envelope envelope, int limit = MaxFrameBytes)
        {
            CheckLimit(limit);
            Validate(envelope);
            int length = envelope.CalculateSize();
            if (length > limit) throw new InvalidDataException("frame exceeds limit");
            byte[] frame = new byte[length + 4];
            frame[0] = (byte)(length >> 24); frame[1] = (byte)(length >> 16);
            frame[2] = (byte)(length >> 8); frame[3] = (byte)length;
            byte[] payload = envelope.ToByteArray();
            Buffer.BlockCopy(payload, 0, frame, 4, length);
            return frame;
        }

        public static Envelope Read(Stream stream, int limit = MaxFrameBytes)
        {
            CheckLimit(limit);
            byte[] header = ReadExact(stream, 4, true);
            if (header == null) return null;
            uint length = ((uint)header[0] << 24) | ((uint)header[1] << 16) | ((uint)header[2] << 8) | header[3];
            if (length == 0 || length > limit) throw new InvalidDataException("invalid frame length");
            Envelope envelope = Envelope.Parser.ParseFrom(ReadExact(stream, (int)length, false));
            Validate(envelope);
            return envelope;
        }

        private static void CheckLimit(int limit)
        {
            if (limit <= 0 || limit > MaxFrameBytes) throw new ArgumentOutOfRangeException("limit");
        }

        private static byte[] ReadExact(Stream stream, int count, bool allowEof)
        {
            byte[] bytes = new byte[count];
            int read = 0;
            while (read < count)
            {
                int received = stream.Read(bytes, read, count - read);
                if (received == 0)
                {
                    if (read == 0 && allowEof) return null;
                    throw new EndOfStreamException("truncated frame");
                }
                read += received;
            }
            return bytes;
        }
    }
}

File: bridges/python/packages/protocol/src/flint_protocol/framing.py
"""Blocking socket framing; callers own scheduling, deadlines and connection closure."""
import struct

from .v1.envelope_pb2 import Envelope

PROTOCOL_VERSION = 1
MAX_FRAME_BYTES = 100 * 1024 * 1024


def validate(envelope):
    if envelope.protocol_version != PROTOCOL_VERSION:
        raise ValueError("unsupported protocol version")
    if not envelope.request_id:
        raise ValueError("missing request_id")
    if envelope.WhichOneof("payload") is None:
        raise ValueError("missing or unsupported payload")


def encode_frame(envelope, limit=MAX_FRAME_BYTES):
    validate(envelope)
    if not 0 < limit <= MAX_FRAME_BYTES:
        raise ValueError("invalid frame limit")
    if envelope.ByteSize() > limit:
        raise ValueError("frame exceeds limit")
    payload = envelope.SerializeToString()
    return struct.pack(">I", len(payload)) + payload


def _read_exact(connection, count, allow_eof=False):
    result = bytearray()
    while len(result) < count:
        chunk = connection.recv(count - len(result))
        if not chunk:
            if not result and allow_eof:
                return None
            raise EOFError("truncated frame")
        result.extend(chunk)
    return bytes(result)


def read_frame(connection, limit=MAX_FRAME_BYTES):
    """Return None only for EOF between complete frames. Close on any error."""
    if not 0 < limit <= MAX_FRAME_BYTES:
        raise ValueError("invalid frame limit")
    header = _read_exact(connection, 4, allow_eof=True)
    if header is None:
        return None
    length = struct.unpack(">I", header)[0]
    if not 0 < length <= limit:
        raise ValueError("invalid frame length")
    envelope = Envelope()
    envelope.ParseFromString(_read_exact(connection, length))
    validate(envelope)
    return envelope


def write_frame(connection, envelope, limit=MAX_FRAME_BYTES):
    connection.sendall(encode_frame(envelope, limit))
